spec/jobs/gpo_expiration_job_spec.rb

```diff
@@ -0,0 +1,168 @@
+        profiles = job.gpo_profiles_that_should_be_expired(as_of: Time.zone.now)
+        expect(profiles.count).to eql(0)
+      end
+    end
+  end
+
+  describe '#perform' do
+    it 'expires the profile' do
+      profile = user_with_one_expired_gpo_profile.reload.gpo_verification_pending_profile
+      freeze_time do
+        expect { job.perform }.to change {
+                                    profile.reload.gpo_verification_expired_at
+                                  }.to eql(Time.zone.now)
+      end
+    end
+
+    it 'clears gpo_verification_pending_at' do
+      profile = user_with_one_expired_gpo_profile.reload.gpo_verification_pending_profile
+      expect { job.perform }.to change { profile.reload.gpo_verification_pending_at }.to eql(nil)
+    end
+
+    it 'logged an analytics event' do
+      job.perform
+      expect(analytics).to have_logged_event(
+        :idv_gpo_expired,
+        user_id: user_with_one_expired_gpo_profile.uuid,
+        user_has_active_profile: false,
+        letters_sent: 1,
+      )
+    end
+
+    context 'when the user has an active profile' do
+      let!(:active_profile) do
+        create(:profile, :active, user: user_with_one_expired_gpo_profile)
+      end
+      it 'includes that information in analytics event' do
+        job.perform
+
+        expect(analytics).to have_logged_event(
+          :idv_gpo_expired,
+          user_id: user_with_one_expired_gpo_profile.uuid,
+          user_has_active_profile: true,
+          letters_sent: 1,
+        )
+      end
+    end
+
+    context 'when the user has multiple codes sent' do
+      let!(:extra_code) do
+        create(
+          :gpo_confirmation_code,
+          profile: user_with_one_expired_gpo_profile.gpo_verification_pending_profile,
+          code_sent_at: expired_timestamp,
+        )
+      end
+
+      it 'we note that in the analytics event' do
+        job.perform
+
+        expect(analytics).to have_logged_event(
+          :idv_gpo_expired,
+          user_id: user_with_one_expired_gpo_profile.uuid,
+          user_has_active_profile: false,
+          letters_sent: 2,
+        )
+      end
+    end
+
+    describe 'limit' do
+      let(:limit) { 3 }
+      before do
+        (0..limit).each do
+          create(
+            :user,
+            :with_pending_gpo_profile,
+            created_at: expired_timestamp,
+          )
+        end
+      end
+      it 'limits the number of records affected' do
+        initial_count = Profile.where.not(gpo_verification_pending_at: nil).count
+
+        job.perform(limit: limit)
+
+        expect(Profile.where.not(gpo_verification_pending_at: nil).count).
+          to eql(initial_count - limit)
+      end
+    end
+  end
+
+  def user_fixture_method_for(profile:)
+    self.methods.
+      map(&:to_s).
+      filter { |method| /^user_with_/.match(method) }.
+      map(&:to_sym).
+      find { |user_fixture_method| profile.user == send(user_fixture_method) }
+  end
```

Comment:
Very fancy! Is there a more straightforward way to do this?

Reply:
I feel like we could build a `users` hash and check that?

```
let(:users) do
  {
    user_with ..... : create(....),
    user_with ...... : create(....),
  }
end
```
and then this spec just checks the keys of that hash?